zap/src/config.rs

```diff
@@ -95,37 +98,85 @@ pub enum Ty<'src> {
+	pub fn size(
+		&self,
+		tydecls: &HashMap<&'src str, TyDecl<'src>>,
+		recursed: &mut HashSet<&'src str>,
+	) -> (usize, Option<usize>) {
 		match self {
-			Self::Num(numty, _) => Some(numty.size()),
-			Self::Color3 => Some(NumTy::U8.size() * 3),
-			Self::Vector3 => Some(NumTy::F32.size() * 3),
-			Self::AlignedCFrame => Some(NumTy::U8.size() + NumTy::F32.size() * 3),
-			Self::CFrame => Some(NumTy::F32.size() * 12),
-			Self::Boolean => Some(1),
-			Self::Opt(ty) => ty.max_size(config, recursed).map(|size| size + 1),
-			Self::Str(len) => len.max().map(|len| len as usize),
-			Self::Buf(len) => len.max().map(|len| len as usize),
-			Self::Arr(ty, range) => range
-				.max()
-				.and_then(|len| ty.max_size(config, recursed).map(|size| size * len as usize)),
-			Self::Map(..) => None,
-			Self::Enum(enum_ty) => enum_ty.max_size(config, recursed),
-			Self::Struct(struct_ty) => struct_ty.max_size(config, recursed),
-			Self::Instance(_) => Some(2),
-			Self::Unknown => None,
+			Self::Num(numty, ..) => (numty.size(), Some(numty.size())),
+
+			Self::Str(len) => {
+				if let Some(exact) = len.exact() {
+					(exact as usize, Some(exact as usize))
+				} else {
+					(
+						len.min().map(|min| (min as usize) + 2).unwrap_or(2),
+						len.max().map(|max| (max as usize) + 2),
+					)
+				}
+			}
+
+			Self::Buf(len) => {
+				if let Some(exact) = len.exact() {
+					(exact as usize, Some(exact as usize))
+				} else {
+					(
+						len.min().map(|min| (min as usize) + 2).unwrap_or(2),
+						len.max().map(|max| (max as usize) + 2),
+					)
+				}
+			}
+
+			Self::Arr(ty, len) => {
+				let (ty_min, ty_max) = ty.size(tydecls, recursed);
+				let len_min = len.min().map(|min| min as usize).unwrap_or(0);
+
+				if let Some(exact) = len.exact() {
+					(ty_min * (exact as usize), ty_max.map(|max| ty_max.unwrap() * max))
+				} else if let Some(len_max) = len.max() {
+					(
+						ty_min * len_min + 2,
+						ty_max.map(|ty_max| ty_max * (len_max as usize) + 2),
+					)
+				} else {
+					(ty_min * len_min + 2, None)
+				}
+			}
+
+			Self::Map(..) => (2, None),
+
+			Self::Opt(ty) => {
+				let (_, ty_max) = ty.size(tydecls, recursed);
+
+				(1, ty_max.map(|ty_max| ty_max + 1))
+			}
+
 			Self::Ref(name) => {
 				if recursed.contains(name) {
-					None
+					// 0 is returned here because all valid recursive types are
+					// bounded and all bounded types have their own min size
+					(0, None)
 				} else {
 					recursed.insert(name);
-					config
-						.tydecls
-						.iter()
-						.find(|tydecl| tydecl.name == *name)
-						.and_then(|tydecl| tydecl.ty.max_size(config, recursed))
+
+					let tydecl = tydecls.get(name).unwrap();
+
+					tydecl.ty.size(tydecls, recursed)
 				}
 			}
+
+			Self::Enum(enum_ty) => enum_ty.size(tydecls, recursed),
+			Self::Struct(struct_ty) => struct_ty.size(tydecls, recursed),
+
+			Self::Instance(_) => (0, Some(0)),
```

Comment:
Should we add a comment explaining why it's `Some(0)` over `None` or do you think it's selfexplainatory?

Reply:
it actually shouldn't be zero, I'll fix it